File: plot_structured_types.py
import os
import pandas as pd
import matplotlib.pyplot as plt

# Create the 'plots' folder if it doesn't exist
if not os.path.exists('plots'):
    os.makedirs('plots')

# Read the data from CSV
data = pd.read_csv('benchmark_results.csv')

# Cast N to string with commas for better readability
data['N_str'] = data['N'].apply(lambda x: f'{x:,}')

# Calculate the average of N over the 3 runs
data_avg = data.groupby(['N', 'Operation', 'Table'])['Time'].mean().reset_index()
data_avg = pd.merge(data_avg, data[['N', 'N_str']].drop_duplicates(), on='N')
print(data_avg)

# Loop through each operation except the scan_value group
for operation, group in data_avg.groupby('Operation'):
    if operation not in ['SCAN_VALUE', 'SCAN_VALUE_1', 'SCAN_VALUE_2']:
        plt.figure(figsize=(10, 8))

        # Extract x and y values, and labels
        x_values = group['N']
        y_values = group['Time']
        x_labels = group['N_str']

        # Plot the lines and markers using pyplot
        for table in group['Table'].unique():
            table_data = group[group['Table'] == table]
            plt.plot(table_data['N_str'], table_data['Time'], marker='o', label=table)

        # Format the plot
        plt.title(f"{operation.capitalize()}, Comparison of Map, Array, Object, 500 to 10m Rows")
        plt.xlabel("Rows")
        plt.ylabel("Time (s)")

        # Add a legend
        plt.legend(title='Table')

        # Save the plot with lowercase operation name
        plt.savefig(f'plots/plot_{operation.lower()}_std.png', dpi=300)
        plt.close()

# Now handle the special case for SCAN_VALUE, SCAN_VALUE_1, and SCAN_VALUE_2
scan_value_data = data_avg[data_avg['Operation'].isin(['SCAN_VALUE', 'SCAN_VALUE_1', 'SCAN_VALUE_2'])]
print(scan_value_data)

# Rename the tables for the legend
scan_value_data['Table'] = scan_value_data['Operation'].replace({
    'SCAN_VALUE': 'array_table',
    'SCAN_VALUE_1': 'map_table_scan_1',
    'SCAN_VALUE_2': 'map_table_scan_2'
})

# Create the combined plot for SCAN_VALUE group
plt.figure(figsize=(8, 6))

# Plot the lines and markers using pyplot
for table in scan_value_data['Table'].unique():
    table_data = scan_value_data[scan_value_data['Table'] == table]
    print(table_data)
    plt.plot(table_data['N_str'], table_data['Time'], marker='o', label=table)

# Format the plot
plt.title(f"Scan, Comparison of Map, Array, Object, 500 to 10m Rows")
plt.xlabel("Rows")
plt.ylabel("Time (s)")

# Add a legend
plt.legend(title='Table')

# Save the plot
plt.savefig('plots/plot_scan_value_std.png', dpi=300)
plt.close()

File: plot_structured_types.R
# Set CRAN mirror
options(repos = c(CRAN = "https://cloud.r-project.org/"))

# Install required packages if they are not installed
packages <- c("dplyr", "ggplot2", "readr", "scales")
new_packages <- packages[!(packages %in% installed.packages()[,"Package"])]
if (length(new_packages)) {
  install.packages(new_packages)
}

# Load required libraries
library(dplyr)
library(ggplot2)
library(readr)
library(scales)

# Create the 'plots' folder if it doesn't exist
if (!dir.exists('plots')) {
  dir.create('plots')
}

# Read the data from CSV
data <- read_csv('benchmark_results.csv')

data <- data %>%
  mutate(Time = as.numeric(Time))

# Calculate the average of N over the 3 runs
data_avg <- data %>%
  group_by(N, Operation, Table) %>%
  summarise(Time = mean(Time), .groups = 'drop')

# Loop through each operation except the SCAN_VALUE group
unique_operations <- unique(data_avg$Operation)

print(data_avg)

for (operation in unique_operations) {
  if (!operation %in% c('SCAN_VALUE', 'SCAN_VALUE_1', 'SCAN_VALUE_2')) {
    # Filter data for the current operation
    group <- data_avg %>% filter(Operation == operation)

    # Create a ggplot for this operation
    p <- ggplot(group, aes(x = N, y = Time, color = Table, group = Table)) +
      geom_line(size = 1) +
      geom_point(size = 2) +
      labs(
        title = paste0(operation, ", Comparison of Map, Array, Object, 500 to 10M Rows"),
        x = "Rows", y = "Time (s)"
      ) +
      theme_bw() +
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.title = element_text(size = 10)
      )

    # Save the plot as PNG file
    ggsave(filename = paste0("plots/plot_", tolower(operation), "_std.png"), plot = p, dpi = 300)
  }
}

# Now handle the special case for SCAN_VALUE, SCAN_VALUE_1, and SCAN_VALUE_2
scan_value_data <- data_avg %>%
  filter(Operation %in% c('SCAN_VALUE', 'SCAN_VALUE_1', 'SCAN_VALUE_2'))

# Rename the tables for the legend
scan_value_data <- scan_value_data %>%
  mutate(Table = recode(Operation,
                        'SCAN_VALUE' = 'array_table',
                        'SCAN_VALUE_1' = 'map_table_scan_1',
                        'SCAN_VALUE_2' = 'map_table_scan_2'))

# Create the combined plot for SCAN_VALUE group
p_scan <- ggplot(scan_value_data, aes(x = N, y = Time, color = Table, group = Table)) +
  geom_line(size = 1) +
  geom_point(size = 2) +
  labs(
    title = "Scan, Comparison of Map, Array, Object, 500 to 10M Rows",
    x = "Rows", y = "Time (s)"
  ) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.title = element_text(size = 10)
  )

# Save the combined scan plot
ggsave(filename = 'plots/plot_scan_value_std.png', plot = p_scan, dpi = 300)
